### PR TITLE
Add new token classification example

### PR DESCRIPTION
**Pull request by sgugger**
# What does this PR do?

This PR adds a new example script leveraging the dataset library, the offsets from the fast tokenizers of the tokenizers library and the Trainer API.

### CODE REVIEW

File: examples/token-classification/run_ner_new.py [hunk 1]
```diff
@@ -0,0 +1,358 @@
...
+            truncation=True,
+            is_split_into_words=True,
+            return_offsets_mapping=True,
+        )
+        offset_mappings = tokenized_inputs.pop("offset_mapping")
+        labels = []
+        for label, offset_mapping in zip(examples[label_column_name], offset_mappings):
+            label_index = 0
+            label_ids = []
+            for offset in offset_mapping:
```
Comment (thomwolf) [inline examples/token-classification/run_ner_new.py:242]: Add a comment to explain that we are setting the label for the first sub-word token of every word.

**Review by thomwolf — APPROVED** [review]
Really cool! And impressive!

Would be great to have a review from @stefan-it as well I think :)

File: examples/token-classification/run_ner_new.py (hunk 1 above)
Comment (sgugger) [reply examples/token-classification/run_ner_new.py:242]: The label is actually repeated for all tokens of the subword. Given your comment, I'm guessing it should not be the case and only set the label for the first subword and then using the label padding index for the rest?

File: examples/token-classification/run_ner_new.py [hunk 2]
```diff
@@ -0,0 +1,358 @@
...
+
+import logging
+import os
+import sys
+from dataclasses import dataclass, field
+from typing import Optional
+
+import numpy as np
+from datasets import load_dataset
+from seqeval.metrics import accuracy_score, f1_score, precision_score, recall_score
```
[the remark this answers is not in the document]
Comment (sgugger) [reply examples/token-classification/run_ner_new.py:28]: Hadn't caught it was also in `datasets`. Yes I do want to use seqeval from `datasets`! This is dependent on [this fix](https://github.com/huggingface/datasets/pull/810) however.

File: examples/token-classification/run_ner_new.py [hunk 3]
```diff
@@ -0,0 +1,361 @@
...
+        # Save predictions
+        output_test_predictions_file = os.path.join(training_args.output_dir, "test_predictions.txt")
+        if trainer.is_world_master():
+            with open(output_test_predictions_file, "w") as writer:
+                for prediction in true_predictions:
+                    writer.write(" ".join(prediction) + "\n")
+
+    return results
+
+    return results
```
Comment (LysandreJik) [inline examples/token-classification/run_ner_new.py:352]: Maybe a single result is enough

**Review by LysandreJik — APPROVED** [review]
LGTM, great explanatory comments!

File: examples/README.md [hunk 4]
```diff
@@ -37,7 +37,7 @@ git checkout tags/v3.4.0
 |---|---|:---:|:---:|:---:|:---:|
 | [**`language-modeling`**](https://github.com/huggingface/transformers/tree/master/examples/language-modeling)       | Raw text        | ✅ | -  | ✅ | [![Open In Colab](https://colab.research.google.com/assets/colab-badge.svg)](https://colab.research.google.com/github/huggingface/blog/blob/master/notebooks/01_how_to_train.ipynb)
 | [**`text-classification`**](https://github.com/huggingface/transformers/tree/master/examples/text-classification)   | GLUE, XNLI      | ✅ | ✅ | ✅ | [![Open In Colab](https://colab.research.google.com/assets/colab-badge.svg)](https://github.com/huggingface/notebooks/blob/master/examples/text_classification.ipynb)
-| [**`token-classification`**](https://github.com/huggingface/transformers/tree/master/examples/token-classification) | CoNLL NER       | ✅ | ✅ | - | -
+| [**`token-classification`**](https://github.com/huggingface/transformers/tree/master/examples/token-classification) | CoNLL NER       | ✅ | ✅ | ✅ | -
```
Comment (LysandreJik) [inline examples/README.md:40]: :heart:

**Review by LysandreJik — APPROVED** [review]

**Comment by stefan-it** [comment]
Hi @sgugger , thanks for these new features :hugs: 

I have a some comments:

* Only training and validation files can be passed via command line argument. The `do_predict` method is already supported, so a new argument for specifiying the test file would be great to have!
* A new version of the `run.sh` script would also be a great feature to start fine-tuning just be executing a shell script. The current `run.sh` script should also get an "_old" suffix then :)

File: examples/token-classification/run_ner.py [hunk 5]
```diff
@@ -73,24 +70,53 @@ class DataTrainingArguments:
...
+    preprocessing_num_workers: Optional[int] = field(
+        default=None,
+        metadata={"help": "The number of processes to use for the preprocessing."},
+    )
+    pad_to_max_length: bool = field(
+        default=False,
         metadata={
-            "help": "The maximum total input sequence length after tokenization. Sequences longer "
-            "than this will be truncated, sequences shorter will be padded."
+            "help": "Whether to pad all samples to `max_seq_length`. "
```
Comment (stefan-it) [inline examples/token-classification/run_ner.py:95]: I think the `max_seq_length` is not used here (and also not specified as an extra command line argument in the new version here).

Later the `pad_to_max_length` argument is specified as:

```python
padding = "max_length" if data_args.pad_to_max_length else False
```

-> so `False` is used and no padding is done :)

File: examples/token-classification/run_pos.sh [hunk 6]
```diff
@@ -21,7 +21,7 @@ export NUM_EPOCHS=3
 export SAVE_STEPS=750
 export SEED=1
 
-python3 run_ner.py \
```
Comment (julien-c) [inline examples/token-classification/run_pos.sh:24]: missing part?

Comment (sgugger) [reply examples/token-classification/run_pos.sh:24]: Oh, very nice catch!

**Pull request merged**